### PR TITLE
testutil/trackpr: add no-issue PRs to github project board

### PR DESCRIPTION
**Pull request by xenowits**
Add PRs with no associated ticket to github project board.

category: misc
ticket: #987

### CODE REVIEW

File: testutil/trackpr/pr/unticketed.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+type PR struct {
+	Title string
+	Body  string
+}
+
+const (
+	prenv = "GITHUB_PR"
+)
```
Comment (dB2510) [inline testutil/trackpr/pr/unticketed.go:35]:
```suggestion
const prEnv = "GITHUB_PR"
```

File: testutil/trackpr/pr/unticketed.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+// Unticketed returns true if the ticket is "none" for the PR and returns false otherwise.
+// It doesn't verify the body and assumes the PR body is already verified before it is merged and closed.
+func Unticketed() (bool, error) {
+	fmt.Println("Verifying charon PR against template")
```
Comment (dB2510) [inline testutil/trackpr/pr/unticketed.go:40]: remove these print statements.
suggest: If logs are needed then use log package

File: testutil/trackpr/pr/unticketed.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package pr provides functions to check and process pull requests.
+//
+//nolint:wrapcheck,cyclop,forbidigo
+package pr
```
Comment (dB2510) [inline testutil/trackpr/pr/unticketed.go:19]: i think we can add unticketed.go in testutil/trackpr package. no need for another sub package for one file

File: testutil/pr/track.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+// getProjectData gets project data given the name of the GitHub organization and the project id.
+func getProjectData(client *githubv4.Client, org string, projectNumber githubv4.Int) (config, error) { //nolint:gocognit
+	variables := map[string]interface{}{
+		"org":    githubv4.String(org),
+		"number": projectNumber,
+	}
+
+	// Note that we query for the first 25 fields.
+	var query struct {
```
Comment (corverroos) [inline testutil/pr/track.go:136]: why is this function local while other JSON types are package level?

File: testutil/pr/track.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+	}
+	if conf.sizeFieldID == nil {
+		return config{}, fmt.Errorf("size field id absent")
+	}
+	if conf.sprintFieldID == nil {
+		return config{}, fmt.Errorf("sprint field id absent")
+	}
+	if conf.currIterationID == nil {
+		return config{}, fmt.Errorf("current iteration id absent")
+	}
```
Comment (corverroos) [inline testutil/pr/track.go:256]: not sure all this validation is required for a simple non-critical tool

File: testutil/pr/track.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+			} `graphql:"item"`
+		} `graphql:"addProjectV2ItemById(input: $input)"`
+	}
+
+	input := AddProjectV2ItemByIdInput{
+		ContentID: prID,
+		ProjectID: projectID,
+	}
+
+	err := client.Mutate(context.Background(), &mutation, input, nil)
```
Comment (corverroos) [inline testutil/pr/track.go:281]: pass context in right from the start

File: testutil/pr/track.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+	unticketed, err := Unticketed()
+	if err != nil {
+		return errors.Wrap(err, "check pr ticket")
+		// log.Fatalf("check pr ticket: %s\n", err.Error())
+	} else if !unticketed {
+		log.Println("ticket exists for this PR")
+		os.Exit(0)
+	}
+
+	ghToken, ok := os.LookupEnv("GH_TOKEN")
```
[the remark this answers is not in the document]
Comment (xenowits) [reply testutil/pr/track.go:95]: sure

File: testutil/pr/track.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+	client := githubv4.NewClient(httpClient)
+
+	// Step 1: Get project data
+	conf, err := getProjectData(client, organization, projectNumber)
+	if err != nil {
+		return errors.Wrap(err, "failed to get project data")
+	}
+
+	// Step 2: Add PR to project board
+	itemID, err := addToProject(client, conf.projectID)
```
[the remark this answers is not in the document]
Comment (xenowits) [reply testutil/pr/track.go:114]: yeah, it adds the pr to the project and the api returns an id of the newly added item.

File: .github/workflows/track-pr.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      GH_TOKEN: ${{ secrets.GH_TOKEN }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.19'
+
+      # Add PR to project board.
+      - name: "Track PR"
+        if: ${{ success() }}
```
Comment (corverroos) [inline .github/workflows/track-pr.yml:21]: what is this "if success"?

File: .github/workflows/track-pr.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+on:
+  pull_request:
+    types: [closed]
+
+jobs:
+  trackpr:
+    runs-on: ubuntu-latest
+    env:
+      GITHUB_PR: ${{ toJSON(github.event.pull_request) }}
+      PR_ID: ${{ github.event.pull_request.node_id }}
```
Comment (corverroos) [inline .github/workflows/track-pr.yml:11]: why add both PR_ID and GITHUB_PR? Seems the same?

File: testutil/pr/trackpr/main.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"log"
+	"os"
+
+	"github.com/obolnetwork/charon/testutil/pr"
+)
+
+func main() {
+	err := pr.Verify()
```
Comment (corverroos) [inline testutil/pr/trackpr/main.go:26]: why verify?

File: testutil/pr/track.go [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+// addProjectV2ItemByIdInput is an input for AddProjectV2ItemById.
+// https://docs.github.com/en/graphql/reference/input-objects#addprojectv2itembyidinput
+type addProjectV2ItemByIdInput struct { //nolint:revive,stylecheck
+	// The ID of the Project to add the item to. (Required.)
+	ProjectID gh.ID `json:"projectId"`
+	// The content id of the item (Issue or PullRequest). (Required.)
+	ContentID gh.ID `json:"contentId"`
+
+	// A unique identifier for the client performing the mutation. (Optional.)
+	ClientMutationID *gh.String `json:"clientMutationId,omitempty"`
```
Comment (corverroos) [inline testutil/pr/track.go:70]: exclude things you do not use, keep it as simple as possible

File: testutil/pr/track.go [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+	ClientMutationID gh.String           `json:"clientMutationId,omitempty"`
+	FieldID          gh.ID               `json:"fieldId"`
+	ItemID           gh.ID               `json:"itemId"`
+	ProjectID        gh.ID               `json:"projectId"`
+	Value            projectV2FieldValue `json:"value"`
+}
+
+// addProjectV2ItemByIdInput is an input for AddProjectV2ItemById.
+// https://docs.github.com/en/graphql/reference/input-objects#addprojectv2itembyidinput
+type addProjectV2ItemByIdInput struct { //nolint:revive,stylecheck
```
Comment (corverroos) [inline testutil/pr/track.go:63]: can you use `gh.UpdatePullRequest`

File: testutil/pr/track.go [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+	Date                 gh.String `json:"date,omitempty"`
+	IterationID          gh.ID     `json:"iterationId,omitempty"`
+	Number               gh.Float  `json:"number,omitempty"`
+	SingleSelectOptionID gh.String `json:"singleSelectOptionId,omitempty"`
+	Text                 gh.String `json:"text,omitempty"`
+}
+
+// updateProjectV2ItemFieldValueInput is an input for UpdateProjectV2ItemFieldValue.
+// https://docs.github.com/en/graphql/reference/input-objects#updateprojectv2itemfieldvalueinput
+type updateProjectV2ItemFieldValueInput struct {
```
Comment (corverroos) [inline testutil/pr/track.go:53]: can't you use `gh.UpdateProjectNextItemFieldInput`?

Comment (xenowits) [reply testutil/pr/track.go:53]: no. the thing is these are "new" objects provided by github api and the library doesn't yet include them. Though, `gh.UpdateProjectNextItemFieldInput` and `updateProjectV2ItemFieldValueInput` are structurally similar, the graphql request throws an error that the input is not a defined type. But, to double confirm, let me try it again.

Comment (xenowits) [reply testutil/pr/track.go:53]: https://github.blog/changelog/2022-06-23-the-new-github-issues-june-23rd-update/#%F0%9F%A4%96-graphql-api-improvements

The `ProjectV2*` stuff isn't there yet in the library

File: testutil/pr/track.go (hunk 13 above)
Comment (xenowits) [reply testutil/pr/track.go:63]: no. the same reason as below. The best replacement would have been `AddProjectNextItemInput` but it'll be deprecated

File: testutil/pr/track.go (hunk 12 above)
Comment (xenowits) [reply testutil/pr/track.go:70]: done

File: testutil/pr/trackpr/main.go (hunk 11 above)
Comment (xenowits) [reply testutil/pr/trackpr/main.go:26]: Not sure if that's important but basically the `Track()` function assumes that the PR is "valid" and doesn't verify it's validity.

This is how i'hv structured it:
`Verify()` simply verifies the PR
`Track()` simply tracks the PR
    - `unticketed()` checks if the PR needs to be tracked.

File: .github/workflows/track-pr.yml (hunk 9 above)
Comment (xenowits) [reply .github/workflows/track-pr.yml:21]: yeah, don't need it as we don't have any previous step. Btw, `success()` returns true when none of the previous steps have failed or been canceled. https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions

File: .github/workflows/track-pr.yml (hunk 10 above)
Comment (xenowits) [reply .github/workflows/track-pr.yml:11]: `PR_ID`: ID generated by GitHub for each PR. Ex: `PR_klEOH3CVWs91nT92`
`GITHUB_PR`: Pull request body for the PR. It contains a `title` and a `body`.

File: testutil/pr/track.go (hunk 4 above)
Comment (xenowits) [reply testutil/pr/track.go]: i refactored all local & package level types to a new file `graphql.go`

File: .github/workflows/track-pr.yml (hunk 10 above)
Comment (corverroos) [reply .github/workflows/track-pr.yml:11]: `GITHUB_PR` also contains the ID, so rather only use that and remove `PR_ID`

File: testutil/pr/trackpr/main.go (hunk 11 above)
Comment (corverroos) [reply testutil/pr/trackpr/main.go]: follow the linux mantra: "do only one thing", so remove `verify`

File: testutil/pr/track.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+		if node.ProjectV2Field.Name == "Size" {
+			conf.sizeFieldID = node.ProjectV2Field.ID
+		}
+
+		// PR status: https://docs.github.com/en/graphql/reference/objects#projectv2singleselectfield
+		if node.ProjectV2SingleSelectField.Name == "Status" {
+			conf.statusFieldID = node.ProjectV2SingleSelectField.ID
+
+			if len(node.ProjectV2SingleSelectField.Options) == 0 {
+				return config{}, fmt.Errorf("status fields absent")
```
Comment (corverroos) [inline testutil/pr/track.go:133]: use `app/errors`

File: testutil/pr/track.go [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+		// sprint iteration: https://docs.github.com/en/graphql/reference/objects#projectv2iterationfielditeration
+		if node.ProjectV2IterationField.Name == "Sprint" {
+			conf.sprintFieldID = node.ProjectV2IterationField.ID
+
+			if len(node.ProjectV2IterationField.Configuration.Iterations) == 0 {
+				return config{}, fmt.Errorf("sprint iterations absent")
+			}
+
+			for _, iter := range node.ProjectV2IterationField.Configuration.Iterations {
+				layout := "2006-01-02"
```
Comment (corverroos) [inline testutil/pr/track.go:152]: can inline

File: testutil/pr/track.go [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+
+				startDate, err := time.Parse(layout, iter.StartDate)
+				if err != nil {
+					return config{}, errors.Wrap(err, "parse date")
+				}
+
+				endDate := startDate.AddDate(0, 0, iter.Duration)
+
+				currSprint := time.Now().Before(endDate)
```
Comment (corverroos) [inline testutil/pr/track.go:161]: also check after startDate

File: testutil/pr/track.go [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+	projectID       gh.ID
+	statusFieldID   gh.ID
+	doneOptionID    gh.String
+	sizeFieldID     gh.ID
+	sprintFieldID   gh.ID
+	currIterationID gh.ID
+}
+
+// Track tracks a PR without a ticket and adds it to the GitHub board.
+func Track(ghToken string) error {
```
Comment (corverroos) [inline testutil/pr/track.go:53]: move all env checks out of this function and pass in as parameters to the function

File: testutil/pr/track.go [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+
+	src := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: ghToken},
+	)
+
+	httpClient := oauth2.NewClient(ctx, src)
+	client := gh.NewClient(httpClient)
+
+	// Step 1: Get project data
+	conf, err := getProjectData(ctx, client, organization, projectNumber)
```
Comment (corverroos) [inline testutil/pr/track.go:73]: suggest passing in globals all the way at the top, not only in the middle

File: testutil/pr/graphql.go [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+	UpdateProjectV2ItemFieldValue struct {
+		ProjectV2Item struct {
+			ID gh.ID `graphql:"id"`
+		} `graphql:"projectV2Item"`
+	} `graphql:"updateProjectV2ItemFieldValue(input: $input)"`
+}
+
+// setStatusMutation sets the status (ex: "Done", "In Progress" etc.) field of the project item.
+// https://docs.github.com/en/graphql/reference/mutations#updateprojectv2itemfieldvalue
+type setStatusMutation struct {
```
[the remark this answers is not in the document]
Comment (xenowits) [reply testutil/pr/graphql.go:112]: yeah, i see, my bad 🙁

File: testutil/pr/track.go (hunk 19 above)
Comment (xenowits) [reply testutil/pr/track.go:73]: i didn't understand this. so, you are suggesting that globals should be globals and we should not pass in global values as parameters, like `getProjectData()` should not have `organization` OR `projectNumber` as params as they are declared globally (package-level0) ?..

File: .github/workflows/track-pr.yml (hunk 10 above)
Comment (xenowits) [reply .github/workflows/track-pr.yml:11]: got it!

File: .github/workflows/track-pr.yml [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+      GITHUB_PR: ${{ toJSON(github.event.pull_request) }}
+      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.19'
+
+      - name: "Track PR"
+        run: go run testutil/trackpr/main.go testutil/trackpr/track.go
```
Comment (corverroos) [inline .github/workflows/track-pr.yml:19]: suggest `go run github.com/obolnetwork/charon/testutil/trackpr`

File: testutil/trackpr/main.go [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
```
Comment (corverroos) [inline testutil/trackpr/main.go:16]: add package comment: `Command trackpr blah blah blah`

File: testutil/trackpr/main.go [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+	log.Println("✅ Success")
+}
+
+func run(ctx context.Context) error {
+	ghToken, ok := os.LookupEnv("GH_TOKEN")
+	if !ok {
+		return errors.New("env not set", z.Str("key", "GH_TOKEN"))
+	}
+
+	p, err := PRFromEnv()
```
Comment (corverroos) [inline testutil/trackpr/main.go:50]: can name `pr` since that package is not there anymore

File: testutil/trackpr/track.go [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	if pr.Title == "" || pr.Body == "" || pr.ID == "" {
+		return PR{}, errors.New("pr field not set")
+	}
+
+	return pr, nil
+}
+
+// track PR without a ticket and add it to the GitHub project board of the organization.
```
Comment (corverroos) [inline testutil/trackpr/track.go:70]: `track adds a PR without a ticket to the provided github project setting size=1, iteration=current, status=done`

**Review by corverroos — APPROVED** [review]

**Pull request merged**